File: python_solutions/1136.parallel-courses.cpp
// Containers
#include <deque>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "conf.d/say.h"

using namespace std;
int dirs[5] = {-1, 0, 1, 0, -1};

// --------------------------------------------------------
class Solution {

public:
  int dfs(int i, vector<vector<int>> &relations, vector<int> &memo) {
    if (memo[i] > -1)
      return memo[i];
    if (memo[i] == -2) // Found a loop
      return 0x3f3f3f3f;
    memo[i] = -2;
    for (int s : relations[i])
      memo[i] = max(memo[i], 1 + dfs(s, relations, memo));
    if (memo[i] == -2)
      memo[i] = 0;
    return memo[i];
  }

  int minimumSemesters(int N, vector<vector<int>> &relations) {
    vector<vector<int>> g(N + 1, vector<int>());
    vector<int> memo(N + 1, -1);
    for (auto &r : relations)
      g[r[0]].push_back(r[1]);

    int res = -1;
    for (int i = 0; i < N; i++) {
      res = max(res, dfs(i, g, memo));
      if (res >= 0x3f3f3f3f)
        return -1;
    }
    return res + 1;
  }
};

int main() {
  Solution s;
  std::vector<vector<int>> relations = {{1, 3}, {2, 3}};
  relations = {{1, 2}, {2, 3}, {3, 1}};
  say(s.minimumSemesters(3, relations));
}

File: python_solutions/1136.parallel-courses.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007


# https://leetcode.com/problems/parallel-courses/description/
# Hard
class Solution:
    def minimumSemesters(self, N: int, relations: List[List[int]]) -> int:
        g = [[] for _ in range(N + 1)]
        memo = [-1] * (N + 1)
        for a, b in relations:
            g[a].append(b)

        def dfs(i):
            if memo[i] > -1:
                return memo[i]
            if memo[i] == -2:
                return 0x3f3f3f3f
            memo[i] = -2
            for j in g[i]:
                memo[i] = max(memo[i], dfs(j) + 1)
            if memo[i] == -2: memo[i] = 0
            return memo[i]

        res = -1
        for i in range(1, N + 1):
            res = max(res, dfs(i))
            if res >= 0x3f3f3f3f: return -1
        return res+1
